students/csimmons/lesson04/trigram.py

```diff
@@ -2,32 +2,76 @@
+# Edited: 12/27/2020 - csimmons
+
+import sys
+import os
+import random
+input = 'sherlock_small.txt'
+welcome_prompt = "\nWelcome to the automatic book generator!"
+
+def read_file(input):
+    all_lines = ''
+    with open(input, 'r') as textfile:
         while True:
             line = textfile.readline()
             if not line:
                 break
-            line = line.replace('\n', '').replace('--', ' ').replace(',','').replace('.', '').replace('(', '').replace(')', '')
-            clean_line = line.split(' ')
-            print(len(clean_line))
-            print(clean_line, '\n')
-            #for i in range(len(clean_line)-2):
-            for i in range(10):
-                pair = clean_line[i:i+2]
-                follower = clean_line[i+2]
-                trigram[tuple(pair)] = follower
-                #print(clean_line)
-                #print(pair)
-                #print(follower)clear
-            print(trigram)
+            all_lines = all_lines + line
+    textfile.close()
+    cleaner(all_lines)
 
-               
+def cleaner(all_lines):
+    clean_lines = all_lines.replace('\n', ' ').replace('--', ' ').replace(',','').replace('.', '').replace('(', '').replace(')', '').replace(':', '')
+    clean_lines = clean_lines.split(' ')
+    # removes any null entries in clean_lines
+    clean_words = list(filter(None, clean_lines)) 
+    create_trigrams(clean_words)
+    
+def create_trigrams(clean_words):
+    trigrams = {}
+    for i in range(len(clean_words)-2):
+        pair = tuple(clean_words[i:i+2])
+        third = clean_words[i+2]
+        if pair in trigrams:
+            trigrams[pair].append(third)
+        else:
+            trigrams[pair] = [third]
+    build_text(trigrams)
 
-        
+def build_text(trigrams):
+    # multiplier used to determine size of text generated for
+    # variable output sizes
+    multiplier = 5
+    num_trigrams = len(trigrams.keys())
+    word_list = []
+    for i in range(num_trigrams * multiplier):
+        key, value = random.choice(list(trigrams.items()))
+        w_one, w_two = key
+        w_three = random.choice(value)
+        word_list.extend([w_one, w_two, w_three])
+    process_text(word_list, trigrams)
+    
+def process_text(word_list, trigrams):
+    # Can add other types of trigram post processing here, including 
+    # re-adding punctuation, capitalization, etc. This function is
+    # easily extendable
+    output_text = ' '.join([str(word) for word in word_list])
+    # prints both trigram dict and randomly generated text.
+    print('\nHere is the trigram dictionary:\n\n', trigrams)
+    print('\nHere is your randomly generated text:\n\n', output_text)
+
+
+def main(input):
+    read_file(input)
+    
 
 if __name__ == '__main__':
-    main(input)
\ No newline at end of file
+    try:
+        input = 'sherlock_small.txt'
+        print(welcome_prompt)
+        main(input)
+    except IndexError:
+        print('Error, program exiting')
+        sys.exit()
```

Comment:
Works great!

Reply:
Yes - trigram is ready for grading.  Will create new PR per your request -